tests/MongoDB.Driver.Core.Tests/Specifications/server-discovery-and-monitoring/TestRunner.cs

```diff
@@ -57,18 +62,101 @@ public void RunTestDefinition(JsonDrivenTestCase testCase)
+                    break;
+
+                }
+                default:
+                    throw new ArgumentException($"Unsupported value of {type} for type");
+            }
+
+            var mockConnection = new Mock<IConnectionHandle>();
+            var isMasterResult = new IsMasterResult(new BsonDocument { { "compressors", new BsonArray() } });
+            var serverVersion = WireVersionHelper.MapWireVersionToServerVersion(maxWireVersion);
+            var buildInfoResult = new BuildInfoResult(new BsonDocument { { "version", serverVersion } });
+            mockConnection.SetupGet(c => c.Description)
+                .Returns(new ConnectionDescription(connectionId, isMasterResult, buildInfoResult));
+            var generation = applicationError.Contains("generation") ? applicationError["generation"].AsInt32 : 0;
+            mockConnection.SetupGet(c => c.Generation).Returns(generation);
+            var when = applicationError["when"].AsString;
+            switch (when)
+            {
+                case "beforeHandshakeCompletes":
+                    server.HandleBeforeHandshakeCompletesException(mockConnection.Object, simulatedException);
+                    break;
+                case "afterHandshakeCompletes":
+
+                    server.HandleChannelException(mockConnection.Object, simulatedException);
+                    // error on application command
+                    break;
+                default:
+                    throw new ArgumentException($"Unsupported value of {when} for when");
+            }
+        }
+
+
         private void ApplyPhase(BsonDocument phase)
         {
-            JsonDrivenHelper.EnsureAllFieldsAreValid(phase, "outcome", "responses");
+            JsonDrivenHelper.EnsureAllFieldsAreValid(phase, "applicationErrors", "description", "outcome", "responses");
+
 
-            var responses = phase["responses"].AsBsonArray;
-            foreach (BsonArray response in responses)
+            if (phase.Contains("responses"))
             {
-                ApplyResponse(response);
+                var responses = phase["responses"].AsBsonArray;
+                foreach (BsonArray response in responses)
+                {
+                    ApplyResponse(response);
+                }
+            }
+
+            if (phase.Contains("applicationErrors"))
+            {
+                var applicationErrors = phase["applicationErrors"].AsBsonArray;
+                foreach (BsonDocument applicationError in applicationErrors)
+                {
+                    ApplyApplicationError(applicationError);
+                }
             }
 
             var outcome = (BsonDocument)phase["outcome"];
-            VerifyOutcome(outcome);
+            var description = phase.Contains("description") ? phase["description"].AsString : "";
+            VerifyOutcome(outcome, description);
```

Comment:
So I've only partially made this change to get some feedback from the rest of the team. Each phase now optionall has a description [(example)](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/tests/errors/stale-topologyVersion-InterruptedAtShutdown.yml#L32). Having this description allow us to pass it to the "Reason" parameter in FluentAssertions, allowing us to effectively see at which phase of the test the test failed.  I've used it a few times in the file below as an example, but haven't changed every fluent assertion to use it because I wanted to see what the rest of the team thought.